### PR TITLE
Organoid segmentations

### PR DESCRIPTION
**Pull request by MikeLippincott**
This PR fixes issues we have been running into with segmentation. There is a lot going on so I will direct to which files are the most important.

0.nuclei_segmentation.py
1.cell_cyto_organoid_segmentation.py
organoid_class_labeling_tool.py
train_random_forrest_binning_model.py
run_chami_75_on_all_images.py
All files in src directory

Thank you!

### CODE REVIEW

File: 2.segment_images/scripts/1.cell_cyto_organoid_segmentation.py [hunk 1]
```diff
@@ -247,24 +231,57 @@
...
 cell_binary_mask[cell_binary_mask > 0] = 1
-# dilate the cell masks slightly
-cell_binary_mask = skimage.morphology.binary_dilation(
-    cell_binary_mask, skimage.morphology.ball(10)
-)
-# convert back to instance mask
+for z in range(cell_binary_mask.shape[0]):
+    cell_binary_mask[z] = scipy.ndimage.binary_fill_holes(
+        cell_binary_mask[z].astype(bool)
+    )
```
Comment (jenna-tomkinson) [inline 2.segment_images/scripts/1.cell_cyto_organoid_segmentation.py]:
```suggestion
# Fill holes in cell masks before generating organoid mask
for z in range(cell_binary_mask.shape[0]):
    cell_binary_mask[z] = scipy.ndimage.binary_fill_holes(
        cell_binary_mask[z].astype(bool)
    )
```
Recommend adding code comments where you can to follow your process :)

File: 2.segment_images/scripts/organoid_class_labeling_tool.py [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+        If True, tries non-blocking display behavior for Jupyter.
+    annotator : str | None
+        Optional annotator name. If None, prompt user.
+    input_mode : str
+        "auto" -> use keypress if backend supports it, else prompt.
+        "keypress" -> force keypress.
+        "prompt" -> force text input prompt.
+    keypress_timeout : float
+        Seconds to wait for keypress before fallback to prompt (0 = no timeout).
+    """
```
Comment (jenna-tomkinson) [inline 2.segment_images/scripts/organoid_class_labeling_tool.py]: Ah I see where my confusion is, please update the parameters to reflect the function above. I missed "label_map" and that answered my question.

File: 3.cellprofiling/scripts/run_chami_75_on_all_images.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+if feature_save_path.exists():
+    print(f"Features already exist at {feature_save_path}, skipping...")
+    sys.exit(0)
+
+
+# In[6]:
+
+
+# Noise Injector transformation
+class SaturationNoiseInjector(nn.Module):
```
[the remark this answers is not in the document]
Comment (jenna-tomkinson) [reply 3.cellprofiling/scripts/run_chami_75_on_all_images.py]: Also, do these functions come from anywhere that would need to be referenced? (e.g., CHAMMI-75 paper?)

File: src/featurization_utils/src/sammed3d_featurizer.py [hunk 4]
```diff
@@ -499,3 +482,64 @@ def call_SAMMed3D_pipeline(
...
+    This function is to be called per patient, well-fov
+    Here we call the SAMMed3D pipeline to extract features for the whole image
+    Parameters
+    ----------
+    image : np.ndarray
+        3D numpy array of the image
+    SAMMed3D_model_path : Optional[str], optional
+        Path to the SAMMed3D model, by default None
+    feature_type : str | List, optional
+        Type of features to extract, by default ["global", "patch", "cls"]
```
[the remark this answers is not in the document]
Comment (jenna-tomkinson) [reply src/featurization_utils/src/sammed3d_featurizer.py]: Also does SAMMED3D generate patches on its own to extract features for?

File: src/segmentation_utils/src/cell_segmentation.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            3D numpy array representing the cytoplasm signal.
+        nuclei_mask : np.ndarray
+            3D numpy array representing the nuclei mask.
+    Returns
+    -------
+        np.ndarray
+            3D numpy array representing the segmented cell mask.
+    """
+    # generate the low frequency elevation map
+    # all morhology types use the same initial elevation map
```
[the remark this answers is not in the document]
Comment (jenna-tomkinson) [reply src/segmentation_utils/src/cell_segmentation.py]: What is an elevation map?

File: src/segmentation_utils/src/organoid_segmentation.py
Comment (jenna-tomkinson) [inline src/segmentation_utils/src/organoid_segmentation.py]: This file looks to only contain edits where functions are commented out, recommend removing the functions or the script if it is not used.

File: 2.segment_images/scripts/1.cell_cyto_organoid_segmentation.py [hunk 6]
```diff
@@ -234,11 +192,37 @@
...
+
 
 # In[12]:
 
 
+# refine the cell masks
+cell_mask = run_post_hoc_refinement(
+    mask_image=cell_mask,
+    sliding_window_context=3,
+)
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 2.segment_images/scripts/1.cell_cyto_organoid_segmentation.py]: updated for clarity!

File: 2.segment_images/scripts/1.cell_cyto_organoid_segmentation.py [hunk 7]
```diff
@@ -247,24 +231,57 @@
...
+
+# In[16]:
+
+
+# nuclei should already have objects removed at the border from the previous notebook,
+# but we can run this again just to be safe
+nuclei_mask = clean_border_objects(nuclei_mask, border_width=25)
+cell_mask = clean_border_objects(cell_mask, border_width=25)
+cytoplasm_mask = clean_border_objects(cytoplasm_mask, border_width=25)
+organoid_mask = clean_border_objects(organoid_mask, border_width=25)
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 2.segment_images/scripts/1.cell_cyto_organoid_segmentation.py]: Ah - good point! I probably should do that. This is added in and updated!

I just find a set of labels and remove any singletons! :)

File: 2.segment_images/scripts/organoid_class_labeling_tool.py [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+                key_pressed = {"key": None}
+
+                def on_key(event):
+                    key_pressed["key"] = event.key
+                    plt.close(fig)
+
+                if in_notebook:
+                    plt.show()
+                    key = input("Enter label for image: ").strip()
+                    plt.close(fig)
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 2.segment_images/scripts/organoid_class_labeling_tool.py]: Added the dict into the prompt here to be more transparent

File: 2.segment_images/scripts/train_random_forrest_binning_model.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+
+# In[5]:
+
+
+labels_save_file = pathlib.Path(
+    "../image_labels/organoid_image_labels.parquet"
+).resolve()
+sammed_features_save_path = pathlib.Path(
+    f"../../3.cellprofiling/results/all_features.parquet"
+).resolve()
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 2.segment_images/scripts/train_random_forrest_binning_model.py]: changed!

File: 2.segment_images/scripts/train_random_forrest_binning_model.py [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+)
+# drop rows with na
+df = df.dropna(subset=["label"])
+
+
+# In[6]:
+
+
+# down sample the training group to be class balanced
+class_counts = df["label"].value_counts()
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 2.segment_images/scripts/train_random_forrest_binning_model.py]: each morphology label in this case

File: 2.segment_images/scripts/train_random_forrest_binning_model.py [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+
+
+all_preds_df.groupby(["patient"]).count()
+
+
+# In[12]:
+
+
+# pick three random images from each predicted class to show
+label_names = {0: "globular", 1: "small/dissociated", 2: "elongated"}
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 2.segment_images/scripts/train_random_forrest_binning_model.py]: I added a docstring to make the goal more clear

File: 3.cellprofiling/scripts/run_chami_75_on_all_images.py [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+def featurize_2D_image_w_chami75(
+    image_tensor: torch.Tensor, model: torch.nn.Module, device: torch.device
+):
+    # Bag of Channels (BoC) - process each channel independently
+    with torch.no_grad():
+        batch_feat = []
+        image_tensor = image_tensor.to(device)
+
+        for c in range(image_tensor.shape[1]):
+            # Extract single channel: (N, C, H, W) -> (N, 1, H, W)
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 3.cellprofiling/scripts/run_chami_75_on_all_images.py]: added comment

File: 3.cellprofiling/scripts/run_chami_75_on_all_images.py [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+    image = read_zstack_image(image_file)
+    # load the middle slice to check if there is anything there
+    mid_slice = image.shape[0] // 2
+    image_mid = image[mid_slice, :, :]
+    images_to_process["patient"].append(patient)
+    images_to_process["well_fov"].append(well_fov)
+    images_to_process["image"].append(image_mid)
+    images_to_process["channel"].append(f"{image_file.stem.split('_')[1]}")
+
+# Convert list of 2D images (H, W) to tensor (B, C, H, W)
```
[the remark this answers is not in the document]
Comment (MikeLippincott) [reply 3.cellprofiling/scripts/run_chami_75_on_all_images.py]: added comments

File: src/segmentation_utils/src/organoid_segmentation.py
Comment (MikeLippincott) [reply src/segmentation_utils/src/organoid_segmentation.py]: removed
